crucible-llvm/src/Lang/Crucible/LLVM/SimpleLoopFixpoint.hs

```diff
@@ -0,0 +1,748 @@
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+
+module Lang.Crucible.LLVM.SimpleLoopFixpoint
+  ( FixpointEntry(..)
+  , simpleLoopFixpoint
+  ) where
+
+import           Control.Lens
+import           Control.Monad.Reader
+import           Control.Monad.State
+import           Control.Monad.Trans.Maybe
+import           Data.Either
+import           Data.Foldable
+import qualified Data.IntMap as IntMap
+import           Data.IORef
+import qualified Data.List as List
+import           Data.Maybe
+import qualified Data.Map as Map
+import           Data.Map (Map)
+import qualified Data.Set as Set
+import qualified System.IO
+import           Numeric.Natural
+
+import qualified Data.BitVector.Sized as BV
+import           Data.Parameterized.Classes
+import qualified Data.Parameterized.Context as Ctx
+import qualified Data.Parameterized.Map as MapF
+import           Data.Parameterized.Map (MapF)
+import           Data.Parameterized.NatRepr
+import           Data.Parameterized.TraversableF
+import           Data.Parameterized.TraversableFC
+
+import qualified What4.Config as W4
+import qualified What4.Interface as W4
+
+import qualified Lang.Crucible.Analysis.Fixpoint.Components as C
+import qualified Lang.Crucible.Backend as C
+import qualified Lang.Crucible.CFG.Core as C
+import qualified Lang.Crucible.Panic as C
+import qualified Lang.Crucible.Simulator.CallFrame as C
+import qualified Lang.Crucible.Simulator.EvalStmt as C
+import qualified Lang.Crucible.Simulator.ExecutionTree as C
+import qualified Lang.Crucible.Simulator.GlobalState as C
+import qualified Lang.Crucible.Simulator.Operations as C
+import qualified Lang.Crucible.Simulator.RegMap as C
+import qualified Lang.Crucible.Simulator as C
+
+import qualified Lang.Crucible.LLVM.Bytes as C
+import qualified Lang.Crucible.LLVM.DataLayout as C
+import qualified Lang.Crucible.LLVM.MemModel as C
+import qualified Lang.Crucible.LLVM.MemModel.MemLog as C hiding (Mem)
+import qualified Lang.Crucible.LLVM.MemModel.Pointer as C
+import qualified Lang.Crucible.LLVM.MemModel.Type as C
+
+
+data FixpointStatus = BeforeFixpoint | ComputeFixpoint | CheckFixpoint | AfterFixpoint
+  deriving (Eq, Ord, Show)
+
+data FixpointEntry sym tp = FixpointEntry
+  { headerValue :: W4.SymExpr sym tp
+  , bodyValue :: W4.SymExpr sym tp
+  }
+
+instance OrdF (W4.SymExpr sym) => OrdF (FixpointEntry sym) where
+  compareF x y = case compareF (headerValue x) (headerValue y) of
+    LTF -> LTF
+    EQF -> compareF (bodyValue x) (bodyValue y)
+    GTF -> GTF
+
+instance OrdF (FixpointEntry sym) => W4.TestEquality (FixpointEntry sym) where
+  testEquality x y = case compareF x y of
+    EQF -> Just Refl
+    _ -> Nothing
+
+instance W4.TestEquality (FixpointEntry sym) => C.EqF (FixpointEntry sym) where
+  eqF x y = isJust $ W4.testEquality x y
```

Comment:
These are slightly scary instances. I suppose the assertion here is that tests for identity are okay, since these definitely don't check for symbolic equality. I think that is worth recording somewhere...

Reply:
@travitch, I don't understand what you think is scary about this? This is just reflecting the syntactic ordering on `SymExpr` into a lexicographic order on the pair (headerValue, bodyValue)

On the other hand, I think it's interesting to ask why we need these instances at all... do they end up as keys in a map or something?

Reply:
I see what they do... I just get nervous around instances like that since someone can see them or accidentally use them without the understanding that they are nonce-identity checks and not semantic checks. I'm just imagining potential frustrating future debugging sessions.

Reply:
Well, if those types aren't exported I guess I have no reason to complain... and the underlying types have those instances anyway. Complaint irrelevant.